modules/fedimint-mint/src/lib.rs

```diff
@@ -371,6 +372,33 @@ impl ModuleGen for MintGen {
     }
 }
 
+impl MintGen {
+    async fn migrate_database(&self, db: Database) -> Result<(), anyhow::Error> {
+        let mut dbtx = db.begin_transaction().await;
+        let ondisk_version = dbtx.get_value(&DatabaseVersionKey).await?;
+        let mint_db_version = if let Some(ondisk_version) = ondisk_version {
+            let current_dbversion = ondisk_version;
+
+            if current_dbversion > DATABASE_VERSION {
+                return Err(anyhow::anyhow!(
+                    "On disk database mint version was higher than the code database version."
+                ));
+            }
+
+            current_dbversion
+        } else {
+            tracing::info!("Mint module no database version found.");
+            dbtx.insert_entry(&DatabaseVersionKey, &DATABASE_VERSION)
+                .await?;
+            dbtx.commit_tx().await?;
+            DATABASE_VERSION
+        };
+
+        tracing::info!("Mint module db version: {}", mint_db_version);
+        Ok(())
+    }
```

Comment:
This code seems very repetitive. I think there should be one function implementing this which is called by all modules in their `init` function. It should take the migration map mentioned by @dpc as argument.


```rust
type MigrationMap = BTreeMap<Version, fn(&mut DatabaseTransaction) -> Box<dyn Future<Output = anyhow::Result<()>>>>

pub fn migrate_db(db: Database, migrations: MigrationMap) -> anyhow::Result<()> {
    …
}
```

This function should run all necessary migrations from that map and update the version key on each step.

There will likely be some lifetime issues around futures capturing a DB tx reference, but see #1423 for how it might be solved.

Reply:
Sure, I can refactor this. I think the `DatabaseTransaction` in the `BTreeMap` doesn't need to be a reference, since we can do each migration as a separate transaction, which would mean we're updating each version atomically instead of doing the entire migration atomically. That might help with the lifetime issues.

Where do you think this map should be populated? In `init`? I tried initializing it statically but ran into problem storing async function inside the `BTreeMap`.

Reply:
Nevermind, got this to work :)